mediactrl: introduce runtime checks for GDK_WINDOWING

varigigi opened the pull request:
For GTK, the current implementation assumes X11 is the only window option.
Introduce runtime checks to manage Wayland too.

Signed-off-by: Pierluigi Passaro <pierluigi.p@variscite.com>

swt2c commented on src/unix/mediactrl.cpp:
```diff
@@ -301,19 +311,32 @@ extern "C" {
+        window_id = (window_id_type)gdk_wayland_window_get_wl_surface(window);
+    }
+#endif
+
```
This looks pretty good to me.  My only suggestion would be to perhaps move this logic for getting the window_id to a separate function and then use that function in both places where you need the window_id.

varigigi replied:
I integrated your suggestion in a function named get_gdk_window_id_from_widget.
Please let me know if you have further suggestions.

varigigi replied:
actually, it looks that similar changes shoudl be applyed to mediactrl_gstplayer.cpp, is it correct?

swt2c replied:
Your changes look good to me, nice work!  Yes, technically mediactrl_gstplayer.cpp needs a similar change for Wayland support.  Thus, it would be nice to share that new function you wrote between both files.  Not sure the best way to do that - it's probably not worth creating a new file for it, but perhaps there's a good place in some common file where it could live as  it may be useful in other places too.  @vadz or @paulcor might have a suggestion.

varigigi replied:
In file mediactrl.cpp, the code use gst_video_overlay_set_window_handle ( => guintptr window_id) for gstreamer 1.x and gst_x_overlay_set_xwindow_id (gulong xwindow_id) for gstremer 0.10.x.
In file mediactrl.cpp, the code use gst_player_video_overlay_video_renderer_set_window_handle and gst_player_video_overlay_video_renderer_new (both using gpointer xwindow_id).
I reworked the code to use exactly the same function get_gdk_window_id_from_widget in both the files: this should make easy export it externally.
Looking forward for further suggestions...

vadz replied:
Yes, it would be definitely better to avoid duplicating this code. The simplest way to do it would be to create a new header under `include/wx/gtk/private` and put it there and make it inline. Please also call it something like `wxGtkGetIdFromWidget()`, i.e. use wx naming convention rather than GTK one to make it clear that it's our function and not anything from GTK.

TIA!

vadz replied:
@varigigi Just to avoid doing redundant work, do you plan to do/are in progress of doing this refactoring already by chance?

varigigi replied:
@vadz, sorry for my absence, it's a crazy week at work.
If not too late, I should be able propose a new patch over the weekend.

vadz replied:
Sure, TIA!

varigigi replied:
@vadz, beyond integrating your suggestions, I've partially reworked the function SetupXOverlay in file src/unix/mediactrl.cpp, trying to make it more readable.
Looking forward for further suggestions.

vadz commented on src/unix/mediactrl.cpp:
```diff
@@ -35,7 +35,13 @@
+    #include "wx/gtk/private/mediactrl.h"
+#endif
+
+#if GST_CHECK_VERSION(1,0,0)
+    typedef guintptr window_id_type;
+#else
+    typedef gulong window_id_type;
 #endif
```
I think this should be moved to `wx/gtk/private/mediactrl.h` too. And ideally it ought to be removed to `wxGtkWindowIdType` or something like this, to make it clear that it's something wx-specific.

varigigi replied:
actually this is not GTK specific: in wxGStreamerMediaBackend::SetupXOverlay(), the windows_id is used even without GTK

vadz commented on include/wx/gtk/private/mediactrl.h:
```diff
@@ -0,0 +1,46 @@
+#ifndef _WX_GTK_PRIVATE_MEDIACTRL_H_
```
Trivial, but could you please add the usual `////...` header comment to the new file for consistency? Just copy it from another file (but use your own name, of course)?

varigigi replied:
added

vadz commented on include/wx/gtk/private/mediactrl.h:
```diff
@@ -0,0 +1,46 @@
+//-----------------------------------------------------------------------------
+// "wxGtkGetIdFromWidget" from widget
+//
+// Get the windows_id performing run-time checks If the window wasn't realized when Load was called, this is the
+// callback for when it is - the purpose of which is to tell
+// GStreamer to play the video in our control
```
Very trivial, but could you please wrap lines to ~80 length, i.e.
```suggestion
// Get the windows_id performing run-time checks If the window wasn't realized
// when Load was called, this is the callback for when it is - the purpose of
// which is to tell GStreamer to play the video in our control.
```

varigigi replied:
done

vadz commented on include/wx/gtk/private/mediactrl.h:
```diff
@@ -0,0 +1,46 @@
+    }
+#endif
+
+    wxSafeShowMessage(_("Fatal Error"), _("wxGtkGetIdFromWidget is only supported on Wayland and X11 currently.  You may be able to\nwork around this by setting environment variable GDK_BACKEND=x11 before\nstarting your program."));
```
The only not completely trivial question I have is whether if we really need this `wxSafeShowMessage()`. It might not be a fatal error (ok, `wxMediaCtrl` is not going to work, but the rest of the program still might). And if we really need it, it would seem to be better to log this message from `wxGStreamerMediaBackend::CreateControl()` using the usual `wxLogError()`, IMO. E.g. do it if `SetupXOverlay()` returns `false`. But to be honest, I'd just not do it at all, I'm not sure the message really adds much.

varigigi replied:
removed ;)

vadz reviewed (commented):
Thanks a lot for the update! This looks almost perfect, but I'd like to understand whether we really need the `wxSafeShowMessage()` call. Of course, it's trivial to just remove it and I can do it myself if you agree with doing it.

varigigi committed:
mediactrl: introduce runtime checks for GDK_WINDOWING

For GTK, the current implementation assumes X11 is the only window option.
Introduce runtime checks to manage Wayland too.

Signed-off-by: Pierluigi Passaro <pierluigi.p@variscite.com>

vadz commented:
Looks great, thanks again! Will merge soon.

varigigi commented:
Thanks for your patient assistance ;)
